### DISCUSSION TITLE
micro-ROS for PlatformIO - Call to action!

### DISCUSSION THREAD

**Discussion by pablogs9** [micro-ROS/micro_ros_platformio#2] [General]
Hi, micro-ROS community!

As you might know, **[PlatformIO](https://registry.platformio.org/) has become a really popular solution for embedded development**, it provides a complete embedded development environment compatible with tons of [boards](https://docs.platformio.org/en/latest/boards/index.html), [frameworks](https://docs.platformio.org/en/latest/frameworks/index.html), and [platforms](https://docs.platformio.org/en/latest/platforms/index.html).

Most of you have been using our [micro-ROS for Arduino](https://github.com/micro-ROS/micro_ros_arduino) library to integrate micro-ROS in the PlatformIO environment, **but we think that we can do it much better**. So, we have been developing a proper solution for integrating the micro-ROS build system in the PlatformIO build system.

[micro-ROS for Arduino](https://github.com/micro-ROS/micro_ros_arduino) provide a precompiled set of libraries due to the lack of advanced features for building the micro-ROS stack in the Arduino IDE/CLI development environment. Meanwhile, PlatformIO allows complex build processes where the micro-ROS library can be compiled just in time for the used platform avoiding cross-compilation compatibility issues. Also, it provides a much more flexible way of configuring the micro-ROS library with [custom packages](https://micro.ros.org/docs/tutorials/advanced/create_new_type/) and/or [custom middleware configurations](https://micro.ros.org/docs/tutorials/advanced/microxrcedds_rmw_configuration/).

**We have reached a semi-functional status** where some platforms are supported but we need your help to provide feedback and comments about your needs as micro-ROS developers. So we are opening a **beta testing period** before releasing this tool on the [PlatformIO registry](https://registry.platformio.org/).

We ask the community to:
- test current available boards and transports
- add / request / recommend new boards and transports
- test the whole functionality and advice new features

Sincerely, the micro-ROS team.

## Technical details

The micro-ROS library for PlatformIO is designed to be modular.  It relies on the [advanced scripting features](https://docs.platformio.org/en/latest/scripting/index.html) of the build system. As you can see in [extra_script.py](https://github.com/micro-ROS/micro_ros_platformio/blob/main/extra_script.py), the library is being built on the fly using the compiler and flags that the environment provides.

In [platform_code folder](https://github.com/micro-ROS/micro_ros_platformio/tree/main/platform_code) we include sources related to the framework. In this case, only `arduino` is implemented. 

This way, [arduino/clock_gettime.cpp](https://github.com/micro-ROS/micro_ros_platformio/blob/main/platform_code/arduino/clock_gettime.cpp) provides a the timing functionality required by micro-ROS using the Arduino API. As you can see, there are much more interesting [frameworks](https://docs.platformio.org/en/latest/frameworks/index.html) such as  [CMSIS](https://docs.platformio.org/en/latest/frameworks/cmsis.html), [FreeRTOS](https://docs.platformio.org/en/latest/frameworks/freertos.html), [libopencm3](https://docs.platformio.org/en/latest/frameworks/libopencm3.html) or [Zephyr RTOS](https://docs.platformio.org/en/latest/frameworks/zephyr.html). That means that those frameworks should be integrated creating `[FRAMEWORK]/clock_gettime.cpp` and adding the corresponding scripting [here](https://github.com/micro-ROS/micro_ros_platformio/blob/65643bb7c9682bd67c81eb8ed97db39a08e2b768/extra_script.py#L113).

Regarding the transports, they are also included in the framework folder. For example, the Serial transport for arduino is implemented using the [`Serial` API of the Arduino framework](https://github.com/micro-ROS/micro_ros_platformio/blob/main/platform_code/arduino/serial/micro_ros_transport.cpp). One more time, the [`extra_script.py`](https://github.com/micro-ROS/micro_ros_platformio/blob/65643bb7c9682bd67c81eb8ed97db39a08e2b768/extra_script.py#L120) takes care of building the required transport. In the same way, by adding folders to this directory tree, more transports can be implemented using different frameworks APIs.

This way you can create a PlaformIO project with the following `.ini`:

```yml
[env:portenta_h7_m7]
platform = ststm32
board = portenta_h7_m7
framework = arduino
microros_transport = serial
microros_distro = galactic
lib_deps =
    https://github.com/micro-ROS/micro_ros_platformio
```

`.meta` files can be added to the enviroment configuration just using `microros_user_meta = [PATH TO YOUR META]` . This will override parameters of the micro-ROS build.

## Tasks to be done

The following tasks should be addressed: 

| Task                       | PR  | Owner          |
| -------------------------- | --- | -------------- |
| Update repo README.md | https://github.com/micro-ROS/micro_ros_platformio/pull/3  | micro-ROS team |
| Add support for Rolling | https://github.com/micro-ROS/micro_ros_platformio/pull/17 | micro-ROS team |
| Implement `extra_packages` | https://github.com/micro-ROS/micro_ros_platformio/pull/12  | micro-ROS team |
| Create example projects | https://github.com/micro-ROS/micro_ros_platformio/pull/16  | micro-ROS team |
| Remove PlatformIO from micro-ROS for Arduino repo | https://github.com/micro-ROS/micro_ros_arduino/pull/957 | micro-ROS team |
| Add clean step for micro-ROS library |  https://github.com/micro-ROS/micro_ros_platformio/pull/15  | micro-ROS team |
| Create a [Renode CI/CD](https://antmicro.com/blog/2022/03/test-driven-development-of-zephyr-micro-ros-with-renode/)  |   |    |

## Board testing

The following use cases shall be tested in hardware: 

| Board               | Platform      | Framework | Transport         | Owner     | Status  |
| ------------------- | ------------- | --------- | ----------------- | --------- | ------- |
| `portenta_h7_m7`    | `ststm32`     | `arduino` | `serial`          | @pablogs9 | Working |
| `teensy41`          | `teensy`      | `arduino` | `serial`          | @pablogs9 | Working |
| `teensy40`          | `teensy`      | `arduino` | `serial`          |           |         |
| `teensy36`          | `teensy`      | `arduino` | `serial`          |           |         |
| `teensy35`          | `teensy`      | `arduino` | `serial`          |           |         |
| `teensy31`          | `teensy`      | `arduino` | `serial`          | @pablogs9 | Working |
| `due`               | `atmelsam`    | `arduino` | `serial`          |           |         |
| `zero`              | `atmelsam`    | `arduino` | `serial`          |           |         |
| `olimex_e407`       | `ststm32`     | `arduino` | `serial`          | @pablogs9 | Working |
| `esp32dev`          | `espressif32` | `arduino` | `serial`          |           |         |
| `nanorp2040connect` | `raspberrypi` | `arduino` | `serial`          | @pablogs9 | Working |
| `teensy41`          | `teensy`      | `arduino` | `native_ethernet` |           |         |
| `nanorp2040connect` | `raspberrypi` | `arduino` | `wifi_nina` | @pablogs9 | Working |
| `portenta_h7_m7`    | `ststm32`     | `arduino` | `wifi`            | @pablogs9 | Working |
| `esp32dev`          | `espressif32` | `arduino` | `wifi`            |           |         |

**Comment by ivankravets** [comment]
Hi @pablogs9,

Thanks for using PlatformIO! 😊  A few comments for integration:

1) `microros_transport = serial`, users will see a warning about the unknowing configuration option. Please use `board_microros.transport  = upload`. Later you can get this data via `env.BoardConfig().get("microros.transport")`

2) `global_env['LIBS'].append(`, `global_env['_CPPDEFFLAGS']`, please don't operate with private SCons scopes. See https://docs.platformio.org/en/latest/projectconf/section_env_build.html#build-flags
    - `global_env.Append(LIBS=['a', 'b', 'c'])`
    - https://github.com/espressif/arduino-esp32/blob/master/tools/platformio-build-esp32.py
    
Please ping me when you apply changes in https://github.com/micro-ROS/micro_ros_platformio/blob/main/extra_script.py

Thanks for your work!

**Reply by pablogs9** [reply]
Hello @ivankravets, thanks a lot for the review.

## First point
With:
```
[env:nanorp2040connect]
platform = raspberrypi
board = nanorp2040connect
framework = arduino
lib_ldf_mode = off
board_microros.transport = serial
lib_deps =
    ../
```

and 

```py
...
# Retrieve the required transport
microros_transport = env.BoardConfig().get("board_microros.transport")
...
```

I get:

```bash
ci git:(main) ✗ pio run -e nanorp2040connect
Processing nanorp2040connect (platform: raspberrypi; board: nanorp2040connect; framework: arduino)
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/raspberrypi/nanorp2040connect.html
PLATFORM: Raspberry Pi RP2040 (1.6.0) > Arduino Nano RP2040 Connect
HARDWARE: RP2040 133MHz, 264KB RAM, 2MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, jlink, raspberrypi-swd)
PACKAGES: 
 - framework-arduino-mbed 3.0.1 
 - tool-rp2040tools 1.0.2 
 - toolchain-gccarmnoneeabi 1.90201.191206 (9.2.1)
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ off, Compatibility ~ soft
All required Python pip packages are installed
KeyError: "Invalid board option 'board_microros.transport'":
  File "/home/pgarrido/.platformio/penv/lib/python3.6/site-packages/platformio/builder/main.py", line 181:
    env.SConscript("$BUILD_SCRIPT")
  File "/home/pgarrido/.platformio/packages/tool-scons/scons-local-4.3.0/SCons/Script/SConscript.py", line 597:
    return _SConscript(self.fs, *files, **subst_kw)
  File "/home/pgarrido/.platformio/packages/tool-scons/scons-local-4.3.0/SCons/Script/SConscript.py", line 285:
    exec(compile(scriptdata, scriptname, 'exec'), call_stack[-1].globals)
  File "/home/pgarrido/.platformio/platforms/raspberrypi/builder/main.py", line 121:
    target_elf = env.BuildProgram()
  File "/home/pgarrido/.platformio/packages/tool-scons/scons-local-4.3.0/SCons/Util.py", line 742:
    return self.method(*nargs, **kwargs)
  File "/home/pgarrido/.platformio/penv/lib/python3.6/site-packages/platformio/builder/tools/platformio.py", line 62:
    env.ProcessProjectDeps()
  File "/home/pgarrido/.platformio/packages/tool-scons/scons-local-4.3.0/SCons/Util.py", line 742:
    return self.method(*nargs, **kwargs)
  File "/home/pgarrido/.platformio/penv/lib/python3.6/site-packages/platformio/builder/tools/platformio.py", line 141:
    project_lib_builder = env.ConfigureProjectLibBuilder()
  File "/home/pgarrido/.platformio/packages/tool-scons/scons-local-4.3.0/SCons/Util.py", line 742:
    return self.method(*nargs, **kwargs)
  File "/home/pgarrido/.platformio/penv/lib/python3.6/site-packages/platformio/builder/tools/piolib.py", line 1104:
    project.install_dependencies()
  File "/home/pgarrido/.platformio/penv/lib/python3.6/site-packages/platformio/builder/tools/piolib.py", line 922:
    if _is_builtin(spec):
  File "/home/pgarrido/.platformio/penv/lib/python3.6/site-packages/platformio/builder/tools/piolib.py", line 914:
    for lb in self.env.GetLibBuilders():
  File "/home/pgarrido/.platformio/packages/tool-scons/scons-local-4.3.0/SCons/Util.py", line 742:
    return self.method(*nargs, **kwargs)
  File "/home/pgarrido/.platformio/penv/lib/python3.6/site-packages/platformio/builder/tools/piolib.py", line 1035:
    lb = LibBuilderFactory.new(env, lib_dir)
  File "/home/pgarrido/.platformio/penv/lib/python3.6/site-packages/platformio/builder/tools/piolib.py", line 61:
    obj = getattr(sys.modules[__name__], clsname)(env, path, verbose=verbose)
  File "/home/pgarrido/.platformio/penv/lib/python3.6/site-packages/platformio/builder/tools/piolib.py", line 149:
    self.process_extra_options()
  File "/home/pgarrido/.platformio/penv/lib/python3.6/site-packages/platformio/builder/tools/piolib.py", line 294:
    exports={"env": self.env, "pio_lib_builder": self},
  File "/home/pgarrido/.platformio/packages/tool-scons/scons-local-4.3.0/SCons/Script/SConscript.py", line 597:
    return _SConscript(self.fs, *files, **subst_kw)
  File "/home/pgarrido/.platformio/packages/tool-scons/scons-local-4.3.0/SCons/Script/SConscript.py", line 285:
    exec(compile(scriptdata, scriptname, 'exec'), call_stack[-1].globals)
  File "/home/pgarrido/dev/micro-ROS/micro_ros_platformio/ci/.pio/libdeps/nanorp2040connect/micro_ros_platformio/extra_script.py", line 49:
    microros_transport = env.BoardConfig().get("board_microros.transport")
  File "/home/pgarrido/.platformio/penv/lib/python3.6/site-packages/platformio/platform/board.py", line 46:
    raise KeyError("Invalid board option '%s'" % path)
============================================================================================= [FAILED] Took 0.87 seconds =============================================================================================
```

## Second point

Can you take a look here: https://github.com/micro-ROS/micro_ros_platformio/pull/10 
Is this PR the correct way of interfacing with SCons?

**Reply by ivankravets** [reply]
> microros_transport = env.BoardConfig().get("board_microros.transport")

It should be `...get("microros.transport")`. The better is to use underscore:

```ini
lib_ldf_mode = off
board_microros_transport = serial
lib_deps =
    ../
```

and Python}

```python
microros_transport = env.BoardConfig().get("microros_transport")
```

**Reply by pablogs9** [reply]
Hello @ivankravets, 

Scons issue is solved here: https://github.com/micro-ROS/micro_ros_platformio/pull/10
The `.ini` issue is addressed here: https://github.com/micro-ROS/micro_ros_platformio/pull/11 But as you can see it does not recognize the parameter. Can you provide some hints?

**Edit**: this is the error https://github.com/micro-ROS/micro_ros_platformio/runs/6059609566?check_suite_focus=true#step:5:194

**Reply by ivankravets** [reply]
I've just tested this:

**platformio.ini**

```ini
[env:myenv]
board_upload.microros_transport = serial
```

**script.py**
```python
print(env.BoardConfig().get("upload.microros_transport", "default_value"))
```

**Reply by pablogs9** [reply]
Same error, this line returns `default_value` as you can see [here](https://github.com/micro-ROS/micro_ros_platformio/runs/6072713636?check_suite_focus=true#step:5:189). If the default value is not provided [it throw an error](https://github.com/micro-ROS/micro_ros_platformio/runs/6072713636?check_suite_focus=true#step:5:190)

**Reply by pablogs9** [reply]
Is there any documentation on that?

**Reply by ivankravets** [reply]
> If the default value is not provided [it throw an error](https://github.com/micro-ROS/micro_ros_platformio/runs/6072713636?check_suite_focus=true#step:5:190)

This is correct behavior. What do you want to achieve? You want to check missing value, just use the next snippet:

```
MISSING = object()
if env.BoardConfig().get("upload.microros_transport", MISSING) == MISSING:
   print("not declared")
```

**Reply by pablogs9** [reply]
But I have defined the parameter...

I want [to write `board_upload.microros_transport = serial` in `platformio.ini`](https://github.com/micro-ROS/micro_ros_platformio/blob/2fb3bb66c59db8afad35ccb82f35e5a1fedecee9/ci/platformio.ini#L7) and then have [`env.BoardConfig().get("upload.microros_transport", "default_value") == "serial"` here](https://github.com/micro-ROS/micro_ros_platformio/blob/2fb3bb66c59db8afad35ccb82f35e5a1fedecee9/extra_script.py#L51).

If `board_upload.microros_transport` is not defined in `platformio.ini` I want to have a default value.

The point is that with the current implementation (as you can see in https://github.com/micro-ROS/micro_ros_platformio/pull/11) I set `board_upload.microros_transport = serial` and when I call  `env.BoardConfig().get("upload.microros_transport", "default_value")` I get `default_value` instead of `serial`.

**Reply by ivankravets** [reply]
Thanks for the report! It was a bug on our side, we have just fixed internally. A temporary solution is to use `global_env.BoardConfig()`. But! Keep using isolated `env.` for the rest part of the code where you extend the build environment related to the library.

**Reply by pablogs9** [reply]
Do you have a PR to track when we can merge the final solution?

**Reply by ivankravets** [reply]
It will be a part of PlatformIO Core 6.0. But! `global_env.BoardConfig()` is a good solution, it will work always. Please use it and don't worry about the upcoming PlatformIO Core 6.0

**Reply by pablogs9** [reply]
Solved in https://github.com/micro-ROS/micro_ros_platformio/pull/11

**Reply by Acuadros95** [reply]
Hi again @ivankravets,

I looks like something changed with latest PlatformIO Core 6.0 release and our custom `CPPDEFINES`:

```c
# Add micro-ROS defines to user application
global_env.Append(CPPDEFINES=[('MICRO_ROS_TRANSPORT_{}_{}'.format(framework.upper(), microros_transport.upper()), 1)])
```

This line adds a define with the framework and transport configured by the user, example: `MICRO_ROS_TRANSPORT_ARDUINO_SERIAL`

We use this symbols to configure different transports on the same source code:

```c
#if defined(MICRO_ROS_TRANSPORT_ARDUINO_SERIAL)
  // ...
#elif defined(MICRO_ROS_TRANSPORT_ARDUINO_NATIVE_ETHERNET)
  // ...
#else
#error "No transport defined"
#endif
```

But with the latest updated on platformIO from [master/get-platformio.py](https://raw.githubusercontent.com/platformio/platformio/master/scripts/get-platformio.py) this defined symbols are missing and we always get to `#error "No transport defined"` in the previous code example.

What are we missing?

**Reply by ivankravets** [reply]
@Acuadros95 , could you provide a simple PlatformIO project (as ZIP archive, attach here) to reproduce this issue? Thanks!

**Reply by Acuadros95** [reply]
@ivankravets Of course, there you go:

[issue_example.zip](https://github.com/micro-ROS/micro_ros_platformio/files/8897104/issue_example.zip)

**Reply by ivankravets** [reply]
> # Add micro-ROS defines to user application

The user's application has its own isolated environment named `projenv`. See example https://docs.platformio.org/en/latest/manifests/library-json/fields/build/extrascript.html

Please note that you will need to use the latest PIO Core Dev (`pio upgrade --dev`). Accessing `projenv` in library extra script was added in PIO Core 6.0.3 that has not been released yet.

Does it work now?

**Reply by Acuadros95** [reply]
Yes!

Is there a way to get the used PIO Core version inside the extrascript?
I would like to keep this working for users with older versions.

**Reply by ivankravets** [reply]
Sorry for the delay, PlaformIO Core 6.1 is out! Please upgrade https://github.com/platformio/platformio-core/releases/tag/v6.1.0

**Reply by Acuadros95** [reply]
Thanks! Upgrading on https://github.com/micro-ROS/micro_ros_platformio/pull/36

**Comment by Acuadros95** [comment]
Just merged `extra_packages` functionality on https://github.com/micro-ROS/micro_ros_platformio/pull/12.
Usage is documented on `README.md` PR https://github.com/micro-ROS/micro_ros_platformio/pull/3.

**Comment by bjsowa** [comment]
I will try to use this library in [my project](https://github.com/LeoRover/leocore_firmware_ros2). It should go fairly easy as it already uses PlatformIO. The project requires support for:

- custom meta file
- custom transport
- custom `clock_gettime`
- extra packages

As all of this is possible now, I'll start trying to switch to this library and report any issues or ask questions when unsure about something.

**Comment by bjsowa** [comment]
What should I set `board_microros_transport` to when I use a custom transport?

**Reply by Acuadros95** [reply]
You have to create a `<framework>/<board_microros_transport>` folder on the `[platform_code](https://github.com/micro-ROS/micro_ros_platformio/tree/main/platform_code)` folder.

For example:
- `platform.ini`:
  ```ini
  framework = arduino
  board_microros_transport = wifi
  ```
- Transport source files: [platform_code/arduino/wifi](https://github.com/micro-ROS/micro_ros_platformio/tree/main/platform_code/arduino/wifi)
- Also, a `MICRO_ROS_TRANSPORT_<FRAMEWORK>_<TRANSPORT>` definition will be available: 
  https://github.com/micro-ROS/micro_ros_platformio/blob/de7a61c7e86fdd0186ed8b7d8ec320994e8ebcbf/ci/src/main.cpp#L3

**Reply by bjsowa** [reply]
I would rather avoid modifying this project to add my custom transport. I currently have the sources for the custom transport under my project's src directory. The library script:
```Python
    # Add platformio library general include path
    global_env.Append(CPPPATH=[
        main_path + "/platform_code",
        main_path + "/platform_code/{}/{}".format(framework, microros_transport)])

    # Add platformio library general to library include path
    env.Append(CPPPATH=[
        main_path + "/platform_code",
        main_path + "/platform_code/{}/{}".format(framework, microros_transport)])
```
will still add additional include paths no matter what I set `board_microros_transport` to. 

Could we, for example, make setting the `board_microros_transport` to `custom` not add the additional include paths?

**Reply by pablogs9** [reply]
That's a good idea. We will take a look at that.

**Reply by pablogs9** [reply]
Working on it here: https://github.com/micro-ROS/micro_ros_platformio/pull/21

**Reply by Acuadros95** [reply]
https://github.com/micro-ROS/micro_ros_platformio/pull/21 merged, the additional include paths will only add the aux `micro_ros_platformio.h` header.

You can check the custom transport section on the repository [README](
https://github.com/micro-ROS/micro_ros_platformio#transport-configuration) and an actual example on our [CI project source](https://github.com/micro-ROS/micro_ros_platformio/blob/main/ci/src/main.cpp), under the `#if defined(MICRO_ROS_TRANSPORT_ARDUINO_CUSTOM)` conditionals.

@bjsowa  Also, we would be happy to add your use case's transport to the repository, feel free to open a PR.

**Comment by Acuadros95** [comment]
Main task completed and merged, we are ready for `Remove PlatformIO from micro-ROS for Arduino repo` step.

**Reply by pablogs9** [reply]
Go ahead

**Reply by Acuadros95** [reply]
Done! https://github.com/micro-ROS/micro_ros_arduino/pull/957
Also added to `micro_ros_setup` docs: https://github.com/micro-ROS/micro_ros_setup/pull/504

**Comment by patrickwasp** [comment]
I've tested the `seeed_xiao` and `mkrzero` boards with serial and they seem to work fine.

**Comment by nhasbun** [comment]
I have been developing my own custom messages within the **extra_packages** folder. This is working well so far, but I find the required re-compilation process of microROS to be a little annoying.

It is very slow and downloads many things behind the scenes, so it is a bit inefficient. Is there a way to force a recompilation of libraries without downloading everything again?

**Reply by pablogs9** [reply]
@Acuadros95

**Reply by Acuadros95** [reply]
> It is very slow and downloads many things behind the scenes, so it is a bit inefficient. Is there a way to force a recompilation of libraries without downloading everything again?

No, this is not possible right now. 

You can modify the build script to avoid deleting the sources, triggering a colcon build over the updated extra_packages. 
But this is not trivial and may take  an excesive ammount of time. 

Maybe you can use a Linux app to test your micro-ROS app,  the  rebuild would be quite faster with this approach.

**Comment by YuriiHordiienko1987** [comment]
Hello, community!
I was spending many time struggling with that error comes from strdup.c library.
The project includes esp32 platform.
Build command:
> pio run -e esp32-espidf

Screen of error.
![Untitled](https://github.com/micro-ROS/micro_ros_platformio/assets/58361495/4e6eeed5-d568-41e5-91a5-590af9f5caa9)
Learning of solutions in www didn't provide me any positive effect. Could it be fixed somehow? Please, take a look.

UPD:
I detected that context of **strdup.c** file which comes after classical installation from [here](https://github.com/micro-ROS/micro_ros_setup/tree/rolling) differs from context provided this platformio repository.
In the classical we have call `rcutils_strndup(str, strlen(str), allocator)`
In PlatformIO we have call `rcutils_strndup(str, SIZE_MAX, allocator)`
There is the root of error behaviour. Could versions of code be aligned in the future?
